### ISSUE TITLE
Allow searching for images

### ISSUE THREAD

**Issue by dcferreira** [neuml/txtai#404]
At the moment the `similar` clause only allows searching for text. It would be useful to extend this to images also.
@davidmezzetti on Slack suggested using something like `similar(image:///PATH)`.

As a workaround for anyone else wanting to search by images, I did notice you can do it right now, but you can't use the SQL syntax. 
That is, you can search the whole index for the closest entry, but can't filter entries out.
This functionality isn't documented on `txtai`, it just works as a side-effect of CLIP.
You can also search for embeddings directly.

For example:
```python
import requests
from sentence_transformers import SentenceTransformer
from PIL import Image
from txtai.embeddings import Embeddings

texts = ["a picture of a cat", "a painting of a dog"]
texts_index = [(i, t, None) for i, t in enumerate(texts)]

embeddings = Embeddings({"method": "sentence-transformers", "path": "sentence-transformers/clip-ViT-B-32", "content": True})
embeddings.index(texts_index)

url = "https://cataas.com/cat"
r = requests.get(url, stream=True)
im = Image.open(r.raw).convert("RGB")

# search image directly
print(embeddings.search(im, 2))

# search embeddings
model = SentenceTransformer('clip-ViT-B-32')
im_emb = model.encode(im)
print(embeddings.search(im_emb, 2))
```
outputs
```text
[{'id': '0', 'text': 'a picture of a cat', 'score': 0.25348278880119324}, {'id': '1', 'text': 'a painting of a dog', 'score': 0.18208511173725128}]
[{'id': '0', 'text': 'a picture of a cat', 'score': 0.25348278880119324}, {'id': '1', 'text': 'a painting of a dog', 'score': 0.18208511173725128}]
```

**Comment by davidmezzetti** [comment]
Thank you for taking the time to submit this. I'll think about options for the similar clause with binary content.

In the meantime, I thought of a way to accomplish this with the current version of txtai.

```python
import glob
import os

from PIL import Image

from txtai.embeddings import Embeddings, Search

def images():
    for path in glob.glob('txtai/*jpg'):
        # Add image object along with image metadata
        image = Image.open(path)
        yield (path, {"object": image, "name": os.path.basename(path)}, None)

class Score:
    def __init__(self):
        self.embeddings, self.search, self.key, self.scores = None, None, None, None

    def __call__(self, indexid, path, limit=3):
        key = hash(f"{path}_{limit}")
        if self.key != key:
            self.key = key
            self.scores = self.query(path, limit)

        return self.scores.get(indexid, 0)

    def query(self, path, limit):
        if not self.search:
            self.search = Search(self.embeddings).search

        return dict(self.search([Image.open(path)], limit)[0])

score = Score()

# Index with content and objects
embeddings = Embeddings({
    "method": "sentence-transformers",
    "path": "sentence-transformers/clip-ViT-B-32",
    "content": True,
    "objects": "image",
    "functions": [score]
})
score.embeddings = embeddings

embeddings.index(images())

embeddings.search("""
  select id, score(indexid, 'txtai/books.jpg', 10) as score from txtai
  where name != 'books.jpg'
""")
```

This code creates a custom SQL function that reads the image and runs a scoring query. It assumes that there is a directory of images in a directory named `txtai` but that can be easily adapted to your use case.

**Comment by dcferreira** [comment]
As a side note, it seems querying for images with that custom `score` function in a SQL query is significantly faster than directly using `embeddings.search(Image.open(PATH))`.

For my case, the SQL query is taking <1ms and searching directly takes ~250ms.

**Comment by dcferreira** [comment]
Yes that was it indeed; I was using the `%timeit` magic command (which runs a line of code multiple times to time it), so of course cache affected it greatly.
Running a single query puts both searches at around the same time. So disregard what I said about the performance gains.

**Comment by davidmezzetti** [comment]
Keeping this issue open, still a good issue to consider.

**Comment by davidmezzetti** [comment]
Resolved with #564
